\documentclass[a4paper,10pt]{article}
\usepackage{hyperref}
\usepackage{setspace}
\usepackage{graphicx}
\usepackage{color}
\usepackage{xcolor}
\usepackage{relsize}
\usepackage{listings}
\usepackage[round]{natbib}
\usepackage{minted}
\usepackage{float}
\hypersetup{colorlinks = true, linkcolor = blue}

\lstset{language=Python}


\usepackage{listings}
\usepackage{makeidx}
\usepackage{titlesec}
\setcounter{tocdepth}{5}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{authblk}
\makeindex{}
%\VignetteEngine{knitr}
%\VignetteIndexEntry{A Markdown Vignette with knitr}
\definecolor{bg}{rgb}{0.95,0.95,0.95}


\title{\textbf{A bioinformatics workflow for detecting signatures of selection in genomic data}}
\date{September 9, 2013}

\author[1,2]{Murray Cadzow}
\author[1,2]{James Boocock}
\author[1,2]{Hoang Tan Nguyen}
\author[2,3]{Phillip Wilcox}
\author[1]{Tony R Merriman}
\author[1]{Michael A Black}
\affil[1]{Department of Biochemistry, University of Otago}
\affil[2]{Department of Mathematics and Statistics, University of Otago}
\affil[3]{Scion Research, Rotorua, New Zealand}
\renewcommand\Authands{ and }
\begin{document}

\maketitle{}
\doublespacing
\tableofcontents





\section{Introduction}
Mik to fill this in
\section{Getting Started}
\subsection{Prerequisites}
The selection pipeline requires some basic libraries and tools are installed on your system. These are python3, Zlib and others.
\subsection{Installation}
To install the package standalone, requiring manual configuration of the config file, run this as root.

\begin{minted}[bgcolor=bg,frame=lines]{bash}
./install.sh --standalone
\end{minted}

The rest of this section will be dedicated to the automatic installation. To perform an automatic installation of the selection pipeline run as root.
\begin{minted}[bgcolor=bg,frame=lines]{bash}
./install.sh
\end{minted}

Installation creates a default config file located in the base directory of the pipeline. Installation adds a program called selection\_pipeline to the system path. To test the program is installed correctly run the following command at a terminal prompt.

\begin{minted}[bgcolor=bg,frame=lines]{bash}
selection_pipeline -h
\end{minted}

The pipeline uses VCF tools, VCF tools requires a perl library to work correctly. Assuming you are using bash as your default shell open your .bashrc file in your home directory and append to the PERL5LIB environment variable.

\begin{minted}[bgcolor=bg,frame=lines]{bash}
# Change the path to match the location of your selection pipeline bin folder.
export PERL5LIB=\${PERL5LIB}:/home/sfk/selection_pipeline/lib/perl5
\end{minted}

To avoid having to change the default config file append the same folder to you PATH environment variable.\\
\begin{minted}[bgcolor=bg,frame=lines]{bash}
export PATH=\${PATH}:/home/sfk/selection_pipeline/bin
\end{minted}

\subsection{Genetic Maps and Impute Haplotypes}
To use the phasing and imputation features of the pipeline requires both genetic map files and haplotype files. For humans these files that conform to the format required for shapeit and impute2 can be found \href{http://mathgen.stats.ox.ac.uk/impute/impute_v2.html#reference}{here}. Download and extract the reference files of your choice and extract them. To use the files with the selection pipeline requires setting a few options in the config file you will use these are as follows an example config file is avaliable in the base directory of the selection pipeline.\\
\begin{minted}[bgcolor=bg,frame=lines]{bash}
[shapeit]
genetic_map_dir = # genetic map file directory #
genetic_map_prefix = # full file name with ? where chromosome number changes #
[impute_executable]
impute_map_dir= # genetic map file directory location #
impute_reference_dir = # impute reference directory location
impute_map_prefix = # the full file name with a ? for where the chromosome number is #
impute_reference_prefix = # The file name with a ? for chromosome name and the extension left out. 
\end{minted}

\section{Tutorial}
\subsection{Selection Signatures at the Lactase Locus}
\subsubsection{Getting the Data}

The lactase gene is located on Chromosome 2 between 136,545,410-136,594,750 positions. For the example we will use a 10 megabase region containing the Lactase gene and the CEU and YRI populations from the 1000 genomes. Usually iHS would be done chromosome wide but in order to demonstrate how to use the pipeline we will use the chromosome 2 region 130,000,000-140,000,000. To download the example dataset enter the command below. 

\begin{minted}[bgcolor=bg]{bash}
wget http://tutorial_file_location.com 
\end{minted}

Extract the data into a new folder

\subsection{Setting up Pipeline Run}
To perform the analysis the first step will is to navigate to the folder you extracted the tutorial data.
\subsection{Population Files}
Population files are required for any cross population comparisions. The commands below will initiate the data generation step.
\begin{minted}[bgcolor=bg,frame=lines]{bash}
multipopultions -p POP1.txt -p POP2.txt -i input.vcf --c defaults.cfg
\end{minted}

The generated folders and current folder have all the data required to perform further selection analysis.

\section{Genome Wide Selection Scans}
The script is designed to work on one chromosome starting with a VCF file containing one or more populations. The pipeline was designed to be applied genome wide in the genome\_wide/ directory we have provided a script that can perform a genome wide analysis. Read the README in this folder for more information.









\bibliographystyle{jss}
\bibliography{refCNVrd2}

\end{document}
