\documentclass[a4paper,10pt]{article}
\usepackage{hyperref}
\usepackage{setspace}
\usepackage{graphicx}
\usepackage{color}
\usepackage{xcolor}
\usepackage{relsize}
\usepackage{listings}
\usepackage[round]{natbib}
\usepackage{minted}
\usepackage{float}
\hypersetup{colorlinks = true, linkcolor = blue}

\lstset{language=Python}


\usepackage{listings}
\usepackage{makeidx}
\usepackage{titlesec}
\setcounter{tocdepth}{5}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{authblk}
\makeindex{}
%\VignetteEngine{knitr}
%\VignetteIndexEntry{A Markdown Vignette with knitr}
\definecolor{bg}{rgb}{0.95,0.95,0.95}


\title{\textbf{A bioinformatics workflow for detecting signatures of selection in genomic data}}
\date{September 9, 2013}

\author[1,2]{Murray Cadzow}
\author[1,2]{James Boocock}
\author[1,2]{Hoang Tan Nguyen}
\author[2,3]{Phillip Wilcox}
\author[1]{Tony R Merriman}
\author[1]{Michael A Black}
\affil[1]{Department of Biochemistry, University of Otago}
\affil[2]{Department of Mathematics and Statistics, University of Otago}
\affil[3]{Scion Research, Rotorua, New Zealand}
\renewcommand\Authands{ and }
\begin{document}

\maketitle{}
\doublespacing
\tableofcontents





\section{Introduction}
Mik to fill this in
\section{Getting Started}
\subsection{Installation}
To install the package standalone will require manual configuration of the config file run this as root.
\begin{minted}[bgcolor=bg]{bash}
./install.sh --standalone
\end{minted}

The rest of this section will be dedicated to the automatic installation. To perform an automatic installation of the selection pipeline run as root.
\begin{minted}[bgcolor=bg]{bash}
./install.sh
\end{minted}

Installation creates a default config file located in the base directory of the pipeline. Installation adds a program to the system path selection\_pipeline. To see the help and test the program is installed correctly run the following command.

\begin{minted}[bgcolor=bg]{bash}
selection_pipeline -h
\end{minted}

\subsection{Selection Signatures at the Lactase Locus}
\subsubsection{Getting the Data}
To perform this tutorial you will be required to install tabix, this package includes the tabix software to install the software on ubuntu requires running the command.
\begin{lstlisting}[language=Bash]
apt-get install tabix
\end{lstlisting}
On other distributions just run as root. 
\begin{lstlisting}[language=Bash]
./install_tabix.sh
\end{lstlisting}
This places the tabix executable in the \emph{bin/} folder. to run tabix type \emph{bin/./tabix} in a terminal from the base directory of the selection pipeline.

The lactase gene is located on Chromosome 2 between 136,545,410-136,594,750 positions. For this example we will use a 10 megabase region containing Lactase and the CEU and YRI populations from the 1000 genomes. Usually iHS would be done chromosome wide but in order to demonstrate how to use thepipeline we will use a small part of chromosome 2 between 130-140 megabases. To download this file from 1000 genomes use the command below you will need to use tabix to download this region.

\begin{lstlisting}[language=Bash]

\end{lstlisting}


\bibliographystyle{jss}
\bibliography{refCNVrd2}

\end{document}
